File: Dia4/Ruby/Animales/Animal.rb
#Clase abstracta

class Animal
    attr_reader :nombre, :edad

    def nombre=(nombre)
        if !nombre.is_a?(String) || nombre.strip.empty?
            raise ArgumentError, "El nombre de un animal debe ser una cadena y no estar vacia"
        end

        @nombre = nombre
    end

    def edad=(edad)
        if !edad.is_a?(Integer) || edad < 0
            raise ArgumentError, "La edad debe ser un entero postivo"
        end

        @edad = edad
    end

    def initialize(nombre, edad)
        if self.class != Animal
            self.nombre = nombre
            self.edad = edad
        else
            raise NotImplementedError, "No se puede construir un Animal (abstracto)"
        end
        
    end

    def hacer_ruido()
        raise NotImplementedError, "Este metodo debe ser implementado en una subclase"
    end
end

File: Dia4/Java/Animales/src/animales/Animal.java
package animales;

public abstract class Animal {
    protected String nombre;
    protected int edad;

    public Animal(String nombre, int edad) {
        setEdad(edad);
        setNombre(nombre);
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {

        if (nombre == null || nombre.isEmpty()) {
            throw new IllegalArgumentException("El nombre de un animal no puede estar vacio");
        }

        this.nombre = nombre;
    }

    public int getEdad() {
        return edad;
    }

    public void setEdad(int edad) {

        if (edad < 0) {
            throw new IllegalArgumentException("La edad de un animal no puede ser negativa");
        }

        this.edad = edad;
    }

    public abstract String hacerRuido();

}
